File: frontend/types/portfolio.ts
export type PortfolioAsset = {
  id: number;
  symbol: string;
  position_size: number;
  average_entry_price: number;
  current_price: number;
  unrealized_pnl: number;
  risk_contribution: number;
  updated_at: string;
};

export type Allocation = {
  id: number;
  target_type: string;
  target_name: string;
  weight: number;
  allocated_amount: number;
  performance_score: number;
  risk_adjusted_return: number;
  correlation_penalty: number;
  stability_score: number;
  drawdown_adjustment: number;
};

export type RebalanceEvent = {
  id: number;
  trigger_reason: string;
  previous_weights: Record<string, number>;
  new_weights: Record<string, number>;
  execution_log: string;
  status: string;
  created_at: string;
};

export type PortfolioMetric = {
  id: number;
  timestamp: string;
  total_equity: number;
  sharpe_ratio: number;
  drawdown: number;
  correlation_risk_score: number;
  exposure_per_asset: Record<string, number>;
  exposure_per_strategy: Record<string, number>;
  volatility_adjusted_return: number;
};

export type RiskSnapshot = {
  id: number;
  timestamp: string;
  scenario_name: string;
  simulated_loss: number;
  limit_status: string;
  metrics_snapshot: Record<string, any>;
};

export type PortfolioCorrelations = {
  symbols: string[];
  matrix: Record<string, Record<string, number>>;
  high_correlation_pairs: [string, string, number][];
  risk_score: number;
  timeframe: string;
  data_available: boolean;
};

export type Portfolio = {
  id: number;
  name: string;
  description: string | null;
  is_active: boolean;
  total_equity: number;
  cash_balance: number;
  peak_equity: number;
  daily_max_risk_pct: number;
  weekly_max_risk_pct: number;
  monthly_max_risk_pct: number;
  created_at: string;
  updated_at: string;
  assets: PortfolioAsset[];
  allocations: Allocation[];
};

export type StrategyPerformance = {
  name: string;
  sharpe_ratio: number;
  win_rate: number;
  profit_factor: number;
  max_drawdown: number;
  stability_score: number;
};

File: backend/app/schemas/portfolio.py
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, Field
from typing import Any, Dict, List, Optional


class PortfolioCreate(BaseModel):
    name: str = Field(default="default")
    description: Optional[str] = None
    initial_balance: Decimal = Field(default=Decimal("10000.0"))
    daily_max_risk_pct: Decimal = Field(default=Decimal("0.02"))
    weekly_max_risk_pct: Decimal = Field(default=Decimal("0.05"))
    monthly_max_risk_pct: Decimal = Field(default=Decimal("0.10"))


class PortfolioAssetOut(BaseModel):
    id: int
    symbol: str
    position_size: Decimal
    average_entry_price: Decimal
    current_price: Decimal
    unrealized_pnl: Decimal
    risk_contribution: Decimal
    updated_at: datetime

    class Config:
        from_attributes = True


class AllocationOut(BaseModel):
    id: int
    target_type: str
    target_name: str
    weight: Decimal
    allocated_amount: Decimal
    performance_score: Decimal
    risk_adjusted_return: Decimal
    correlation_penalty: Decimal
    stability_score: Decimal
    drawdown_adjustment: Decimal

    class Config:
        from_attributes = True


class RebalanceOut(BaseModel):
    id: int
    trigger_reason: str
    previous_weights: Dict[str, float]
    new_weights: Dict[str, float]
    execution_log: str
    status: str
    created_at: datetime

    class Config:
        from_attributes = True


class PortfolioMetricOut(BaseModel):
    id: int
    timestamp: datetime
    total_equity: Decimal
    sharpe_ratio: Decimal
    drawdown: Decimal
    correlation_risk_score: Decimal
    exposure_per_asset: Dict[str, float]
    exposure_per_strategy: Dict[str, float]
    volatility_adjusted_return: Decimal

    class Config:
        from_attributes = True


class RiskSnapshotOut(BaseModel):
    id: int
    timestamp: datetime
    scenario_name: str
    simulated_loss: Decimal
    limit_status: str
    metrics_snapshot: Dict[str, Any] = {}

    class Config:
        from_attributes = True
        arbitrary_types_allowed = True


class PortfolioOut(BaseModel):
    id: int
    name: str
    description: Optional[str]
    is_active: bool
    total_equity: Decimal
    cash_balance: Decimal
    peak_equity: Decimal
    daily_max_risk_pct: Decimal
    weekly_max_risk_pct: Decimal
    monthly_max_risk_pct: Decimal
    created_at: datetime
    updated_at: datetime
    assets: List[PortfolioAssetOut] = []
    allocations: List[AllocationOut] = []

    class Config:
        from_attributes = True
